feat: add readthedocs javascript search

ReenigneArcher opened the pull request:
This PR adds a custom readthedocs search extension using javascript. I found the default doxygen search is quite terrible, and doesn't seem to actually search any markdown files (or barely searches them)... which is not great when you have user documentation.

I have added this for my own projects (here: https://github.com/LizardByte/doxyconfig/pull/4), and figured I could share it back here for others to benefit from. A live preview is available here: https://docs.lizardbyte.dev/projects/doxyconfig/latest/search.html?query=doxygen

This is what it looks like.
![image](https://github.com/user-attachments/assets/dc32cb6a-e45c-4f1f-9735-c702d44ee6d9)

![image](https://github.com/user-attachments/assets/25ad208b-428a-42e3-8888-aa39d4864cb9)


<details>
  <summary>Old version</summary>

   ![image](https://github.com/user-attachments/assets/e7a7d9de-27dd-499b-99fd-bca805129eca)

   ![image](https://github.com/user-attachments/assets/9b87cf22-eebc-4c28-a970-0951180ed1cc)
</details>

The styling is somewhat based on [Furo](https://github.com/pradyunsg/furo), this is how that looks in one of my projects.
![image](https://github.com/user-attachments/assets/6abaf736-dfe1-42cf-b6b4-39f9b5f7e3dc)


TODO:
- [x] improve the appearance/layout of the anchor urls
- [x] possibly add live searching (This is above my capability for now, maybe it could be part of a later PR).

ReenigneArcher committed:
feat: add readthedocs javascript search

ReenigneArcher commented:
A user of mine reported there is an issue with the search results page on mobile. I don't think it's anything related to my code, but possibly a doxygen bug?

When the user clicks the search bar, the keyboard pops up, but then immediately closes... and there is no way to enter a search term. I know doxygen generates search pages slightly differently depending on the variable search config options. Seems to work perfectly fine on desktop though.

This code is in the original doxygen `search.js`, so maybe something to do with this.

```javascript
function SearchBox(name, resultsPath, extension) {
  this.DOMSearchField = () => document.getElementById("MSearchField");
  this.DOMSearchBox = () => document.getElementById("MSearchBox");
  this.OnSearchFieldFocus = isActive => {
    if (isActive) {
      this.DOMSearchBox().className = 'MSearchBoxActive';
    } else {
      this.DOMSearchBox().className = 'MSearchBoxInactive';
    }
  }
}
```

ReenigneArcher commented on doxygen-awesome-readthedocs-search.js:
```diff
@@ -0,0 +1,172 @@
+      // readthedocs metadata
+      // TODO: how to handle defaults? ... only matters when this is outside of readthedocs
+      let projectSlug = DoxygenAwesomeReadtheDocsSearch.getMetaValue("readthedocs-project-slug") || "doxygen-awesome-css";
+      let projectVersion = DoxygenAwesomeReadtheDocsSearch.getMetaValue("readthedocs-version") || "latest";
```
```suggestion
      let projectVersion = DoxygenAwesomeReadtheDocsSearch.getMetaValue("readthedocs-version-slug") || "latest";
```

ReenigneArcher committed:
fix(rtd): update version variable

jothepro commented on package.json:
```diff
@@ -28,7 +28,8 @@
   "config": {},
-  "dependencies": {},
-  "devDependencies": {},
+  "devDependencies": {
+    "jquery": "^3.7.1"
+  },
```
was `jquery` intentionally listed as a `devDependency` here? Currently the `package.json` is only used to allow for installation of the custom CSS. Can you explain your development workflow that requires the dependency to be added here?

ReenigneArcher replied:
Yes, otherwise the IDE complains about undefined names, such as `ajax` if I recall.

jothepro reviewed (commented):
Cool! The readthedocs search seems to work a lot better indeed! I think adding this as an extension to the theme would be really valuable!

When starting the search with the readthedocs addon widget at the bottom right, a really nice popover with live search is opened. IMO that's a really nice user experience! Do you think it would be possible to use the searchbar in the upper right just as a link that opens this popover?

ReenigneArcher commented:
> When starting the search with the readthedocs addon widget at the bottom right, a really nice popover with live search is opened. IMO that's a really nice user experience! Do you think it would be possible to use the searchbar in the upper right just as a link that opens this popover?

I'm not sure I fully understand the ask, but if you're just asking if the live search will work (search as you type)... it's possible, I just don't know how to implement it.

ReenigneArcher committed:
fix(rtd): Add live RTD search dropdown and focus fixes

Introduce a live search dropdown and various robustness improvements for ReadTheDocs search integration. Changes:
- Add CSS for live search dropdown (#RTDLiveResults) styling and item/snippet formatting.
- Use globalThis.location.hostname and expose searchFor on globalThis to avoid scope issues.
- Wrap global SearchBox to preserve behavior while preventing CloseResultsWindow from blurring an actively focused search field.
- Add MutationObservers to re-attach listeners and re-focus the search input when the DOM is mutated (fixes mobile/OSK behavior where input is recreated).
- Implement debounced live search (_attachLiveSearch, _runLiveSearch), AJAX-based live result rendering (_showLiveResults), positioning, keyboard navigation, and hide logic.
- Refactor paginated search fetch into fetchResults with a ctx object and appendResultItem helper for consistent DOM updates.
- Make getReadTheDocsDefaultVersion return a Promise and resolve to 'latest' on error (handles PR numeric versions).

These changes improve UX for live searching, keyboard support, and resilience when menu code recreates the search field.

ReenigneArcher committed:
refactor(rtd): Replace jQuery calls with native DOM/fetch

Migrate away from jQuery for DOM access and AJAX: use document.getElementById / querySelector and textContent/appendChild instead of jQuery selectors and methods; replace $.ajax calls with fetch + response.ok checks and promise handlers; replace $.each with Array.forEach; update search field assignment to use .value. Remove jquery from package.json devDependencies. This modernizes the code and removes an unnecessary runtime dependency.

ReenigneArcher commented:
Decided to revisit this PR with assistance from Claude.

- Fixed issue where on screen keyboard would disappear when searching on a mobile device
- Add live search
- Fixed several sonar issues
- Replaced jquery/ajax with native fetch

jothepro commented on package.json:
```diff
@@ -28,7 +28,5 @@
   },
   "license": "MIT",
   "config": {},
-  "dependencies": {},
-  "devDependencies": {},
```
Why was this changed? It doesn't seem to be related to the PR?

ReenigneArcher replied:
I'll revert. Must have been when I had jquery in here as a dev dependency and then removed it.

jothepro commented on doxygen-awesome-readthedocs-search.css:
```diff
@@ -0,0 +1,87 @@
+}
+
+#RTDLiveResults .rtd-live-item:hover,
+#RTDLiveResults .rtd-live-item:focus {
+    background: var(--menu-focus-background);
+    color: var(--menu-focus-foreground);
+}
+
+#RTDLiveResults .rtd-live-title {
+    display: block;
+    font-weight: bold;
+    font-size: var(--navigation-font-size);
+}
+
+#RTDLiveResults .rtd-live-snippet {
+    display: block;
+    font-size: calc(var(--navigation-font-size) * 0.9);
+    color: var(--page-secondary-foreground-color);
+    white-space: nowrap;
+    overflow: hidden;
+    text-overflow: ellipsis;
+    max-width: 400px;
+}
```
It seems like the `:hover` and `:focus` overrides do not work. I suggest to set the color `--menu-focus-foreground` to both `.rtd-live-title` and `.rtd-live-snippet` explicitly to fix this:


```suggestion
#RTDLiveResults .rtd-live-item:hover,
#RTDLiveResults .rtd-live-item:focus {
    background: var(--menu-focus-background);
}

#RTDLiveResults .rtd-live-title {
    display: block;
    font-weight: bold;
    font-size: var(--navigation-font-size);
}

#RTDLiveResults .rtd-live-snippet {
    display: block;
    font-size: calc(var(--navigation-font-size) * 0.9);
    color: var(--page-secondary-foreground-color);
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    max-width: 400px;
}

#RTDLiveResults .rtd-live-item:hover .rtd-live-title, #RTDLiveResults .rtd-live-item:hover .rtd-live-snippet, #RTDLiveResults .rtd-live-item:focus .rtd-live-title, #RTDLiveResults .rtd-live-item:focus .rtd-live-snippet {
    color: var(--menu-focus-foreground);
}
```

## Before:

<img width="400" alt="Image" src="https://github.com/user-attachments/assets/752e1b7f-f9aa-4761-adf1-1ac469f5c7ea" />
<img width="400" alt="Image" src="https://github.com/user-attachments/assets/a9114ede-adf0-49db-aed2-7b2f2dd58a76" />

## After:

<img width="400" alt="Image" src="https://github.com/user-attachments/assets/12c029be-98f8-4e6d-8b52-b9f87b48af65" />
<img width="400" alt="Image" src="https://github.com/user-attachments/assets/d0472987-604f-4e05-8906-634ff670cb23" />

jothepro commented on doxygen-awesome-readthedocs-search.css:
```diff
@@ -0,0 +1,87 @@
+
+#RTDLiveResults .rtd-live-list li {
+    margin: 0;
+}
+
+#RTDLiveResults .rtd-live-item {
+    display: block;
+    padding: var(--spacing-small) var(--spacing-medium);
+    color: var(--page-foreground-color);
+    text-decoration: none;
+    outline: none;
+}
```
I'd prefer if there was some spacing on the focused items, to stay consistent with the spacing in the drop-down menus:


```suggestion
#RTDLiveResults .rtd-live-item {
    display: block;
    padding: var(--spacing-small) var(--spacing-medium);
    margin: var(--spacing-small);
    border-radius: var(--border-radius-small);
    color: var(--page-foreground-color);
    text-decoration: none;
    outline: none;
}
```

Result:

<img width="500" alt="Image" src="https://github.com/user-attachments/assets/03ec315d-d193-441b-b1e8-95b1a1240078" />

jothepro commented on doxygen-awesome-readthedocs-search.js:
```diff
@@ -0,0 +1,392 @@
+
+  static _positionLiveResults(dropdown) {
+    const field = document.getElementById('MSearchField');
+    const box = document.getElementById('MSearchBox');
+    const anchor = box || field;
+    if (!anchor) return;
+    const rect = anchor.getBoundingClientRect();
+    const scrollY = window.scrollY;
+    const viewportWidth = document.documentElement.clientWidth;
+    dropdown.style.top = `${rect.bottom + scrollY}px`;
+    dropdown.style.left = '';
+    dropdown.style.right = `${viewportWidth - rect.right}px`;
+    dropdown.style.minWidth = `${Math.max(rect.width, 260)}px`;
```
I assume that this will not work with the sidebar only variant of the theme, because the element is always aligned to the right edge of the search box. The sidebar theme would require a left-align though, otherwise the element will not be fully visible.

I've crafted some screenshots on how this would likely look like with the current alignment, and how the layout should be IMO:

## Right Aligned

<img width="500" alt="Image" src="https://github.com/user-attachments/assets/a4fbaf9f-f6ff-43eb-b1a4-34453ad1803f" />

## Left Aligned

<img width="550" alt="Image" src="https://github.com/user-attachments/assets/d34f1423-3f8b-4589-ad9b-7a00c5e37a20" />

I'm guessing an easy fix would be to add an init parameter that allows to configure the alignment?

```js
DoxygenAwesomeReadtheDocsSearch.init('leftAlign')
```

Pls correct me if my assumptions are not correct.

jothepro commented on doxygen-awesome-readthedocs-search.css:
```diff
@@ -0,0 +1,87 @@
+    margin-bottom: 10px;
+}
+
+/* Live search dropdown */
+#RTDLiveResults {
+    display: none;
+    position: absolute;
+    z-index: 10000;
+    left: unset;
+    background: var(--page-background-color);
+    border: 1px solid var(--separator-color);
+    border-radius: var(--border-radius-large);
+    box-shadow: var(--box-shadow);
+    max-height: 420px;
+    overflow-y: auto;
+}
```
The element does not scale well on small screens. I'd suggest to add an `@media` selector for mobile devices that fixes the width of the RTDLiveResults div:


```suggestion
#RTDLiveResults {
    display: none;
    position: absolute;
    z-index: 10000;
    left: unset;
    background: var(--page-background-color);
    border: 1px solid var(--separator-color);
    border-radius: var(--border-radius-large);
    box-shadow: var(--box-shadow);
    max-height: 420px;
    overflow-y: auto;
}

@media screen and (max-width: 767px) {
    #RTDLiveResults {
        left: var(--spacing-medium);
        right: var(--spacing-medium);
    }
}
```

`767px` is the Breakpoint from desktop to mobile used elsewhere in the code as well.

## Result

<img width="400" alt="Image" src="https://github.com/user-attachments/assets/a2c636a1-2775-4e3e-9501-c4c45fcfd55b" />

jothepro commented on doxygen-awesome-readthedocs-search.css:
```diff
@@ -0,0 +1,87 @@
+}
+
+/* Live search dropdown */
+#RTDLiveResults {
```
It's just a personal preference but I think a small margin between the search box and the RTDLiveResults looks a lot better:

```suggestion
#RTDLiveResults {
    margin-top: var(--spacing-small);
```

## Result

<img width="645" height="252" alt="Image" src="https://github.com/user-attachments/assets/3c1f58ec-e69e-4c00-aeb5-84c061bce2d8" />

jothepro commented on doxygen-awesome-readthedocs-search.js:
```diff
@@ -0,0 +1,392 @@
+    const rect = anchor.getBoundingClientRect();
+    const scrollY = window.scrollY;
+    const viewportWidth = document.documentElement.clientWidth;
+    dropdown.style.top = `${rect.bottom + scrollY}px`;
```
Very minor bug that I'd consider tolerable because I'd consider the input a rase edge case:

When resizing the viewport, the position of the dropdown is not recalculated. This occurs when switching from desktop to mobile layout and when the navigation-items move under the logo:

https://github.com/user-attachments/assets/42ced27b-e93c-458e-b6e7-819a62d3ae8a

jothepro commented on doxygen-awesome-readthedocs-search.js:
```diff
@@ -0,0 +1,392 @@
+            e.preventDefault();
+            focused.click();
+          }
+        }
```
I'm not sure why but the keyboard navigation is broken for me:

It is possible to focus the first item in the list pressing <kbd>↓</kbd> on the keyboard. But the focus is stuck then and cannot be moved to other items using either <kbd>↓</kbd> or <kbd>↑</kbd>. It is also not possible to exit out of the search by pressing <kbd>Esc</kbd>.

I've tried to visualize the problem in this video (Keyboard inputs are displayed in the lower left)

https://github.com/user-attachments/assets/6df115ea-2bfd-4f3f-9354-3b7bed6fce66

jothepro reviewed (changes requested):
This looks very promising already, thanks for the good work!

ReenigneArcher committed:
chore: restore dep groups in package.json

ReenigneArcher committed:
fix: apply css styling suggestions from code review

Co-Authored-By: jothepro <21294002+jothepro@users.noreply.github.com>

ReenigneArcher committed:
fix: Add left-aligned live search and position fixes

Support left-aligned live search variant and improve live-results behavior. Docs: add example showing initialization with leftAlign for the sidebar-only theme. JS: add alignment state, allow init('leftAlign'), prevent duplicate live-search attachment, move keydown handling to a shared _handleLiveSearchKeydown, add dropdown keydown listener, restore focus on Escape, update positioning logic for left/right alignment and mobile, use requestAnimationFrame to schedule position updates, and listen for window resize to recompute placement. Overall fixes keyboard, focus and positioning edge cases and reduces redundant event attachment.

ReenigneArcher committed:
chore: add readthedocs config and build scripts
